nikolai_chernyshov/graph_generator.cpp

```diff
@@ -26,43 +32,117 @@ const VertexId& get_random_vertex_id(const std::vector<VertexId>& vertex_ids) {
+
+  green_thread.join();
+  yellow_thread.join();
+  red_thread.join();
 
   return graph;
 }
 
-void GraphGenerator::add_gray_edges(Graph& graph) const {
-  double proba_step = 1.0 / (double)params_.depth;
-
-  graph.add_vertex();
-  for (int depth = 0; depth < params_.depth; depth++) {
-    // создал копию, чтобы не итерироваться по изменяемому массиву
-    // т.к. при создании серых граней у новых вершин пересчитывается глубина
-    const auto vertex_ids = graph.get_vertex_ids_in_depth(depth);
-
-    for (const auto& vertex_id : vertex_ids) {
-      for (int generate_try_num = 0;
-           generate_try_num < params_.new_vertices_num; generate_try_num++) {
-        if (get_random_boolean(1 - (double)depth * proba_step)) {
-          const auto new_vertex_id = graph.add_vertex();
-          graph.add_edge(vertex_id, new_vertex_id);
+void GraphGenerator::generate_gray_edges(Graph& graph) const {
+  using JobCallback = std::function<void()>;
+  auto jobs = std::list<JobCallback>();
+
+  std::mutex jobs_mutex;
+  std::atomic<int> finished_jobs_num = 0;
+  std::atomic<bool> should_terminate = false;
+
+  const VertexId root_vertex_id = graph.add_vertex();
+
+  for (int i = 0; i < params_.new_vertices_num; i++) {
+    jobs.push_back(
+        [&graph, &jobs_mutex, &root_vertex_id, &finished_jobs_num, this]() {
+          generate_gray_branch(graph, 0, root_vertex_id, jobs_mutex);
+          finished_jobs_num++;
+        });
+  }
+
+  auto worker = [&should_terminate, &jobs_mutex, &jobs]() {
+    while (true) {
+      if (should_terminate) {
+        return;
+      }
+      const auto job_optional = [&jobs,
+                                 &jobs_mutex]() -> std::optional<JobCallback> {
+        const std::lock_guard lock(jobs_mutex);
+        if (!jobs.empty()) {
+          const auto job = jobs.front();
+          jobs.pop_front();
+          return job;
         }
+        return std::nullopt;
+      }();
+      if (job_optional.has_value()) {
+        const auto& job = job_optional.value();
+        job();
       }
     }
+  };
+
+  const auto threads_count =
+      std::min(MAX_THREADS_COUNT, params_.new_vertices_num);
+  auto threads = std::vector<std::thread>();
+  threads.reserve(threads_count);
+
+  for (int i = 0; i < threads_count; i++) {
+    threads.push_back(std::thread(worker));
+  }
+
+  while (finished_jobs_num < params_.new_vertices_num) {
+  }
+
+  should_terminate = true;
+  for (auto& thread : threads) {
+    thread.join();
   }
 }
 
-void GraphGenerator::add_green_edges(Graph& graph) const {
+void GraphGenerator::generate_gray_branch(Graph& graph,
+                                          int depth,
+                                          const VertexId& parent_vertex_id,
+                                          std::mutex& mutex) const {
+  if (depth == params_.depth) {
+    return;
+  }
+
+  if (!get_random_boolean(1 - (double)depth / (double)params_.depth)) {

```

Comment:
Непонятно, что здесь происходит...
Давай лучше так:
```suggestion
  const double probability = 1 - (double)depth / (double)params_.depth); 
  if (!get_random_boolean(probability) {
```